src/FreeAge/common/modifications.hpp

```diff
@@ -0,0 +1,247 @@
+  Civilization = 0,
+
+  // Units
+  AllUnits,
+  AllMilitaryUnits,
+  UnitByType,
+  UnitsByArmor,
+
+  // Buildings
+  AllBuildings,
+  AllBuildingsExceptDefences,
+  BuildingByType,
+  BuildingsByArmor,
+
+  // Techonolies
+  AllTechnologies, // except ages
+  TechnologyByType,
+};
+
+/// A set of conditions for a unit, building, technology or civilization that can be used
+/// to specify inderectly a collection of them.
+///
+/// Example: This following filter matches only with units that have the Cavalry armor class.
+///    ObjectFilter filter = ObjectFilter::UnitsByArmor(DamageType::Cavalry);
+///
+/// An object filter can be created only with one of the factory methods.
+struct ObjectFilter {
+
+  static const ObjectFilter Civilization() {
+    return ObjectFilter(ObjectFilterType::Civilization, 0);
+  }
+  static const ObjectFilter AllUnits() {
+    return ObjectFilter(ObjectFilterType::AllUnits, 0);
+  }
+  static const ObjectFilter AllMilitaryUnits() {
+    return ObjectFilter(ObjectFilterType::AllMilitaryUnits, 0);
+  }
+  static const ObjectFilter UnitByType(UnitType unitType) {
+    return ObjectFilter(ObjectFilterType::UnitByType, static_cast<int>(unitType));
+  }
+  static const ObjectFilter UnitsByArmor(DamageType damageType) {
+    return ObjectFilter(ObjectFilterType::UnitsByArmor, static_cast<int>(damageType));
+  }
+  static const ObjectFilter AllBuildings() {
+    return ObjectFilter(ObjectFilterType::AllBuildings, 0);
+  }
+  static const ObjectFilter AllBuildingsExceptDefences() {
+    return ObjectFilter(ObjectFilterType::AllBuildingsExceptDefences, 0);
+  }
+  static const ObjectFilter BuildingsByType(BuildingType buildingType) {
+    return ObjectFilter(ObjectFilterType::BuildingByType, static_cast<int>(buildingType));
+  }
+  static const ObjectFilter BuildingsByArmor(DamageType damageType) {
+    return ObjectFilter(ObjectFilterType::BuildingsByArmor, static_cast<int>(damageType));
+  }
+  static const ObjectFilter AllTechnologies() {
+    return ObjectFilter(ObjectFilterType::AllTechnologies, 0);
+  }
+  static const ObjectFilter TechnologyByType(Technology technology) {
+    return ObjectFilter(ObjectFilterType::TechnologyByType, static_cast<int>(technology));
+  }
+
+  bool MatchesCivilization() const;
+  bool MatchesUnits() const;
+  bool MatchesBuildings() const;
+  bool MatchesTechnologies() const;
+
+  bool MatchesUnit(const Player& player, const UnitType& unitType) const;
+  bool MatchesBuilding(const Player& player, const BuildingType& buildingType) const;
+  bool MatchesTechnology(const Player& player, const Technology& technology) const;
+
+ private:
+
+  ObjectFilter(ObjectFilterType type, int data)
+      : type(type),
+        data(data) {};
+
+  const ObjectFilterType type;
+
+  // TODO: use union ?
+  const int data;
+
+};
+
+/// A immutable pair of a filter and a modification. The modification should be able to apply to
+/// the objects that are matched by the filter.
+struct TargetedModification {
```

Comment:
In the future, there could be a TargetedModification to string method which would allow to have the list of modifications that the technology provides be generated instead of having a need for a description of the technology.